crud/src/main/java/com/redhat/lightblue/assoc/ep/ExecutionPlan.java

```diff
@@ -0,0 +1,474 @@
+        QueryPlanNode[] allNodes=qp.getAllNodes();
+        for(QueryPlanNode node:allNodes) {
+            if(node.getData()!=null)
+                node.getData().getConjuncts().stream().forEach(c->qfi.addAll(c.getFieldInfo()));
+            for(QueryPlanNode source:node.getSources()) {
+                QueryPlanData edgeData=qp.getEdgeData(source,node);
+                if(edgeData!=null)
+                    edgeData.getConjuncts().stream().forEach(c->qfi.addAll(c.getFieldInfo()));
+            }
+        }
+        return qfi;
+    }
+        
+        
+
+    /**
+     * Writes a projection to include the given set of fields
+     */
+    public static Projection writeProjection(Collection<Path> fields) {
+        List<Projection> list=new ArrayList<>(fields.size());
+        for(Path p:fields)
+            list.add(new FieldProjection(p,true,false));
+        return list.size()==1?list.get(0):new ProjectionList(list);
+    }
+
+    /**
+     * Returns the included fields of an entity based on query
+     * information. This is to be used for search plan nodes only
+     */
+    public static Set<Path> getIncludedFieldsOfEntityForSearch(ExecutionBlock block,
+                                                               List<QueryFieldInfo> fieldInfo) {
+        CompositeMetadata md=block.getMetadata();
+        Set<Path> ret=new HashSet<>();
+        for(QueryFieldInfo fi:fieldInfo) {
+        	if(md==fi.getFieldEntity()&&!(fi.getFieldMd() instanceof ArrayField ||
+                                          fi.getFieldMd() instanceof ObjectField)) {
+        		ret.add(fi.getEntityRelativeFieldNameWithContext());
+        	}
+        }
+        // Add the identities
+        for(Path p:block.getIdentityFields())
+            ret.add(p);
+        return ret;
+    }
+
+    /**
+     * Returns the fields included based on the requested projection and reference field projection
+     */
+    public static Set<Path> getIncludedFieldsOfEntityForProjection(ExecutionBlock block,
+                                                                   CompositeMetadata root,
+                                                                   Projection requestProjection) {
+        Set<Path> fields=new HashSet<>();
+        // What is the path prefix for this entity?
+        CompositeMetadata md=block.getMetadata();
+        Path entityPath=md.getEntityPath();
+        Path globalPrefix=entityPath.isEmpty()?Path.EMPTY:new Path(entityPath,Path.ANYPATH);
+        // entityPath is either empty, meaning this is the root, or a path to a reference field
+        // If entityPath is not empty, then find the projection in the reference field
+        ResolvedReferenceField reference=block.getReference();
+        Projection localProjection;
+        if(reference!=null) {
+            localProjection=reference.getReferenceField().getProjection();
+        } else {
+            localProjection=null;
```

Comment:
I usually avoid double-assignments like that.

On Wed, Mar 30, 2016 at 10:33 AM, Brandon Vulaj <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/ep/ExecutionPlan.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57920247>
> :
>
> > +    public static Set<Path> getIncludedFieldsOfEntityForProjection(ExecutionBlock block,
> > +                                                                   CompositeMetadata root,
> > +                                                                   Projection requestProjection) {
> > +        Set<Path> fields=new HashSet<>();
> > +        // What is the path prefix for this entity?
> > +        CompositeMetadata md=block.getMetadata();
> > +        Path entityPath=md.getEntityPath();
> > +        Path globalPrefix=entityPath.isEmpty()?Path.EMPTY:new Path(entityPath,Path.ANYPATH);
> > +        // entityPath is either empty, meaning this is the root, or a path to a reference field
> > +        // If entityPath is not empty, then find the projection in the reference field
> > +        ResolvedReferenceField reference=block.getReference();
> > +        Projection localProjection;
> > +        if(reference!=null) {
> > +            localProjection=reference.getReferenceField().getProjection();
> > +        } else {
> > +            localProjection=null;
>
> @bserdar <https://github.com/bserdar> do we need the else statement to
> set it to null? Just set to null on declaration and if reference!=null set
> it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/d72d06258b9b58a53294c8eb4a994145df771c4c#r57920247>
>